apps/emqx_web_hook/test/emqx_actions_test_utils.erl

```diff
@@ -0,0 +1,206 @@
+    % ct:pal("======> Create rules:>~p~nwith options: ~p~n", [RuleData, RulesOptions]),
+    0 = get(<<"code">>, RuleData),
+    RuleID = maps:get(<<"id">>, get(<<"data">>, RuleData)),
+    RuleID.
+
+delete_resource(ResourceID) ->
+    {ok, _} = request_api(delete, api_path(["resources", binary_to_list(ResourceID)]), auth_header()).
+
+delete_rule(RuleID) ->
+    {ok, _} = request_api(delete, api_path(["rules", binary_to_list(RuleID)]), auth_header()).
+
+create_resource_failed(SourceOptions) ->
+    {ok, R} = request_api(post, api_path(["resources"]), [], auth_header(), SourceOptions),
+    400 = get(<<"code">>, R),
+    ct:pal("======> Create resource failed test OK\n").
+
+create_rules_failed(RulesOptions) ->
+    {ok, RuleData} = request_api(post, api_path(["rules"]), [], auth_header(), RulesOptions),
+    400 = get(<<"code">>, RuleData),
+    ct:pal("======> Create rules failed test OK\n").
+
+get_rule_matched(RuleID) ->
+    {ok, Data} = request_api(get, api_path(["rules",binary_to_list(RuleID)]), auth_header()),
+    [RuleMetric | _ ] = maps:get(<<"metrics">>, get(<<"data">>, Data)),
+    [Action | _ ] = maps:get(<<"actions">>, get(<<"data">>, Data)),
+    [ActionMetric | _ ] = maps:get(<<"metrics">>, Action, 0),
+    #{matched => maps:get(<<"matched">>, RuleMetric),
+      success => maps:get(<<"success">>, ActionMetric),
+      failed => maps:get(<<"failed">>, ActionMetric)}.
+
+update_action_options(Opts = #{<<"params">> := BaseParams}, Params) ->
+    Opts#{<<"params">> => maps:merge(BaseParams, Params)}.
+
+make_action_options(ActName, Params) ->
+    #{<<"name">> => ActName,
+      <<"params">> => Params}.
+
+clean_start_emulator(EmulatorName) ->
+    process_flag(trap_exit, true),
+    case EmulatorName:start_link() of
+        {error, {already_started, _}} ->
+            EmulatorName:stop(),
+            {ok, _} = EmulatorName:start_link(),
+            ok;
+        {ok, _} -> ok
+    end.
+
+total_records_count(SqlList) ->
+    total_records_count(SqlList, sql).
+
+total_records_count(SqlList, SqlType) ->
+    io:format("---- list: ~p", [SqlList]),
+    lists:foldl(fun(Sql, Acc) ->
+            Acc + records_count(Sql, SqlType)
+        end, 0, SqlList).
+
+records_count(Sql, sql) ->
+    case emqx_rule_actions_utils:split_insert_sql(str(Sql)) of
+        {ok, {_, Values}} ->
+            length(re:split(Values, "\\),", [{return, list}]));
+        {error, not_insert_sql} ->
+            1
+    end;
+
+records_count(Sql, cassandra) ->
+    case re:split(Sql, "((?i)values)",[{return,binary}]) of
+        [_InsertPart | ParamsPart] ->
+            length([V || V <- ParamsPart, V =/= <<"values">>]);
+        _ -> error(not_cassandra_sql)
+    end;
+
+records_count([" values (" | Sql], dolphindb) ->
+    Data = [Elm || Elm <- hd(hd(Sql)), Elm =/= "[", Elm =/= "]", Elm =/= ","],
+    length(Data);
+records_count(Sql, dolphindb) when is_binary(Sql) ->
+    1;
+
+records_count("INSERT ALL" ++ Sql, oracle) ->
```

Comment:
Nit: add trailing space after ALL to make sure the caller doesn't need to care about leading spaces. Same for the other clause